shared/src/price_estimate.rs

```diff
@@ -268,6 +268,69 @@ impl UniswapPriceEstimator {
     }
 }
 
+pub struct MultiAmmPriceEstimator {
+    amm_estimators: Vec<UniswapPriceEstimator>,
+}
+
+#[async_trait::async_trait]
+impl PriceEstimating for MultiAmmPriceEstimator {
+    async fn estimate_price(
+        &self,
+        sell_token: H160,
+        buy_token: H160,
+        amount: U256,
+        kind: OrderKind,
+    ) -> Result<BigRational> {
+        // Return min price estimate as best case.
+        let estimates = join_all(self.amm_estimators.iter().map(|estimator| async move {
+            estimator
+                .estimate_price(sell_token, buy_token, amount, kind)
+                .await
+        }))
+        .await;
+
+        let mut valid_estimates = vec![];
+        for estimate in estimates {
+            if let Ok(price) = estimate {
+                valid_estimates.push(price)
+            }
+        }
+
+        match valid_estimates.len() {
+            0 => Err(anyhow!("Failed to estimate price on all estimators")),
+            _ => Ok(valid_estimates.iter().min().unwrap().clone()),
+        }
```

Comment:
Here is an equivalent shorter version. Can probably do something like this for the version below too.
```suggestion
        estimates
            .into_iter()
            .filter_map(Result::ok)
            .min()
            .ok_or_else(|| anyhow!("Failed to estimate prices on all estimators"))
```